File: fr.sivrit.svn.helper.scala/src/fr/sivrit/svn/helper/scala/svn/SvnClient.scala
package fr.sivrit.svn.helper.scala.svn
import org.tigris.subversion.svnclientadapter.SVNUrl
import java.io.File
import fr.sivrit.svn.helper.Preferences
import org.tigris.subversion.svnclientadapter.ISVNInfo
import org.tigris.subversion.svnclientadapter.SVNNodeKind
import org.tigris.subversion.svnclientadapter.ISVNDirEntry

class SvnClient {
  def fetch(url: SVNUrl): SvnNode = {
    val svn: SvnAdapter = SvnAdapter.borrow

    try {
      val info: ISVNInfo = svn.getInfo(url)

      val version: Long = info.getLastChangedRevision.getNumber

      return fetch(url, version, info.getNodeKind() == SVNNodeKind.DIR)
    } finally {
      SvnAdapter.release(svn);
    }
  }

  def fetch(url: SVNUrl, version: Long, isDir: Boolean): SvnNode =
    {
      val svn: SvnAdapter = SvnAdapter.borrow()
      try {
        if (isDir) {
          val content: Array[ISVNDirEntry] = svn.getList(url)
          var entries: Set[SvnFolderEntry] = Set.empty
          for (isvnDirEntry <- content) {
            entries += new SvnFolderEntry(isvnDirEntry)
          }

          return new SvnDir(url, version, entries);
        } else {
          val content: String = svn.getStringContent(url)
          return new SvnFile(url, version, content);
        }
      } finally {
        SvnAdapter.release(svn);
      }
    }
}

object SvnClient {
  def createClient(): SvnClient = {
    val location = Preferences.getCacheFolder
    if (location == null) {
      new SvnClient();
    } else {
      val cacheDir = if (location.endsWith(File.separator)) location else (location
        + File.separator);
      return new CachedSvnClient(cacheDir);
    }
  }
}

File: fr.sivrit.svn.helper.java/src/fr/sivrit/svn/helper/java/svn/SvnClient.java
package fr.sivrit.svn.helper.java.svn;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;

import org.tigris.subversion.subclipse.core.SVNException;
import org.tigris.subversion.svnclientadapter.ISVNDirEntry;
import org.tigris.subversion.svnclientadapter.ISVNInfo;
import org.tigris.subversion.svnclientadapter.SVNClientException;
import org.tigris.subversion.svnclientadapter.SVNNodeKind;
import org.tigris.subversion.svnclientadapter.SVNUrl;

import fr.sivrit.svn.helper.Preferences;

public class SvnClient {
    public static SvnClient createClient() {
        final String location = Preferences.getCacheFolder();
        if (location == null) {
            return new SvnClient();
        } else {
            final String cacheDir = location.endsWith(File.separator) ? location : location
                    + File.separator;
            return new CachedSvnClient(cacheDir);
        }
    }

    protected SvnClient() {
        super();
    }

    public SvnNode fetch(final SVNUrl url) throws SVNClientException, SVNException {
        final SvnAdapter svn = SvnAdapter.create();

        final ISVNInfo info = svn.getInfo(url);
        final long version = info.getLastChangedRevision().getNumber();

        return fetch(url, version, info.getNodeKind() == SVNNodeKind.DIR);
    }

    public SvnNode fetch(final SVNUrl url, final long version, final boolean isDir)
            throws SVNClientException, SVNException {
        final SvnAdapter svn = SvnAdapter.create();

        if (isDir) {
            final ISVNDirEntry[] content = svn.getList(url);
            final Collection<SvnFolderEntry> entries = new ArrayList<SvnFolderEntry>();
            for (final ISVNDirEntry isvnDirEntry : content) {
                entries.add(new SvnFolderEntry(isvnDirEntry));
            }

            return new SvnNode(url, version, entries);
        } else {
            final String content = svn.getStringContent(url);
            return new SvnNode(url, version, content);
        }
    }
}
